### PR TITLE
Add HandleExt trait on unix for sending signals

### PR DESCRIPTION
**Pull request by faern**
Follow up to https://github.com/oconnor663/shared_child.rs/pull/13

I have not added tests or verified that this works as intended yet. But I wanted to get the PR in place so the solution and design could be discussed.

### CODE REVIEW

**Comment by faern** [comment]
Now I added two initial tests testing some basic stuff. The question is how the signals should be sent to the more advanced chains of commands. Should a signal be sent to all processes in a pipe for example? How can I test for this? I only get one `Output` out etc.

**Comment by oconnor663** [comment]
I gave it a quick look just now, and I agree with this approach, where we walk the tree and signal all the children individually. Apologies for how complicated these types are already, especially `ThenHandle`, with so many sensitive details about when we take and release which locks. As part of reviewing this properly, I'll think some more about whether we can get rid of any of that complexity.

Here are a few details I'll want to think about:

- When you're signaling `pipe` and `then`, what happens when signaling one side of the expression returns an error? Do we still signal the other side? `wait` and `kill` try to guarantee that cleanup/signaling happens all over the tree, even when one part returns errors. That's why we store the right side of a pipe inside a `Result<>`, for example, but unfortunately it's a major source of complexity.
  - Speaking of which, what errors are even possible when signaling? I guess if we have bugs in our own waiting code, we could run into the "PID does not exist" condition, though hopefully that's impossible if our code is correct. Maybe it's possible the parent process could drop its permissions, and then we'd get a permissions error when we try to signal our own children? Or maybe we could just get errors when the system is totally out of RAM? The details might not matter too much, but I'm curious.
- There's an extra subtlety with `then`, which we have to think about in `kill`. The question is what happens if you signal while the left side is running, before the right side has started. Does the right side start after that? In the most common case it won't, because the left side will return non-zero. But there are exceptions to that. The left side might be `unchecked`, so that its return code gets ignored. Or the left side might've _just_ exited, and we might be in that tiny window between exit and when the parent's `wait` will return zero, when signaling presumably has no effect. So anyway, `ThenHandle::kill` has logic for this, to make sure we never start the right side after a kill. Unfortunately, it's not at all clear to me whether we can do the same thing for `send_signal`. That would require hardcoding which signals have the "intent to kill" I think? That sounds too magical to me. Is it enough to just document these weird conditions? Is the fact that the user _isn't_ just calling `kill` enough for us to say they _don't_ have the "intent to kill"? I'm really not sure yet.

**Comment by oconnor663** [comment]
By the way, the only alternative I know of to signaling all the children individually, is to put all the children in their own process group and then signal that group. That seems like how Unix really wanted us to do it, and it has the benefit that grandchildren processes that we don't know about would usually get signaled too. Unfortunately it has the major downside of preventing a Ctrl-C in the parent process from propagating normally to the children (because I think terminals rely on process groups for that magic), which we can't fix without installing a SIGINT handler in the parent to manually forward the signal to child groups. I think that's what shells do, but as a library we can't do it :(

**Comment by faern** [comment]
Sorry for being idle. I have been busy with other stuff. I have read your answers, but I have not had time to investigate further or come up with some solution. I agree we have to be careful what should be signaled in advanced execution trees.

**Comment by oconnor663** [comment]
It's also possible we'll decide that the `then` operator is too obscure to be worth supporting, given how many complexities it introduces. I don't want to cut it out, but I should think about it.

**Comment by oconnor663** [comment]
Two more thoughts:

- This might be an argument that we need a (maybe Unix-only) `terminate` method or something like that after all? To capture the intent.
- I need to figure out what bash does. If I run `sleep 1; sleep 1` and hit ctrl-c, is there a race condition somewhere in there? I don't know.

**Comment by faern** [comment]
Running the command you wrote will never spawn the second sleep. I don't know how it is implemented. It feels reasonable that it does not have race conditions. I guess Bash is smart enough to know that ctrl+c should abort and flag that somewhere before sending SIGTERM to the first process. Then it just never starts the second process based on the flag it set.

**Comment by faern** [comment]
Maybe the thread waiting for the first child to finish can simply check the `ExitStatus` of it? The `ExitStatus` should tell if the process died by a signal. If so it should probably never start the second child. Or maybe not for a subset of valid signals or something.

**Comment by faern** [comment]
I pushed some updates. First commit updates the logic to actually always send the signal to all processes even if a previous signal failed.

The second commit adds a dedicated `terminate` method that works around the problem with `send_signal` and `then` expressions you were concerned about. `terminate` works like `kill`, except sending SIGTERM instead.

As long as the documentation explains the various strange cases I'm happy with this type of functionality. But raise any concerns you have and we'll see if we can improve it in any way.

**Comment by faern** [comment]
I agree that hardcoding which signals has the intent to kill is way too magical. I think we should keep `send_signal` fairly simple, just sending the signal to all currently running subprocesses. Instead of doing something fancy in code we can outline the edge cases and caveats for it in its documentation.

File: src/test.rs [hunk 1]
```diff
@@ -15,7 +15,7 @@ use std::process::Command;
 use std::str;
 use std::sync::{Arc, Once, ONCE_INIT};
 
-fn path_to_exe(name: &str) -> PathBuf {
+pub fn path_to_exe(name: &str) -> PathBuf {
```
Comment (oconnor663) [inline src/test.rs]: Why this change?

Comment (faern) [inline src/test.rs]: To be able to use it in my tests in `unix.rs` where I do `use test::*;`
Just now when I looked at that again I realized I could also make `true_cmd` public and make use of that.

**Comment by faern** [comment]
Rebased on latest master and squashed some of the small fix commits into the commits they were fixing.

**Comment by oconnor663** [comment]
Sorry for the slow feedback on this one.

The more I think about this, the more I think `then` is not pulling its weight, given the amount of complexity it brings along for so many other features like this. I think people are much more likely to use a simple, reliable signals implementation than they are to use `then`. What would you think if we just removed it? (cc @olson-sean-k)

**Comment by faern** [comment]
@oconnor663 Personally I just need `duct` for quite simple single child processes and the possibility to control the io streams of that child as well as to kill it in different ways (important to be able to SIGTERM). So for my usage of this library removing `then` would be completely fine.

A user of `duct` should be able to replicate this functionality quite easily anyway, so better keep the library simpler then?

**Comment by faern** [comment]
So do you want to merge this anyway or rip out `then` first? Or change the signal/terminate implementation for `then` into a nop or panic or something and then merge it?

**Comment by oconnor663** [comment]
I think if we don't have to worry about weirdo `then` race conditions, then we probably don't need the separate `terminate` method? If that sounds right to you, maybe we could take that out and then merge this. Otherwise, let me know what you think? (My thinking is similar to https://github.com/oconnor663/shared_child.rs/pull/13.)

**Comment by faern** [comment]
I think a dedicated `terminate` still has a reason to exist. If the documentation for it clearly states that what it does is to simply send a `SIGTERM`. I don't know other peoples' use cases for duct, but for me the only signal I ever need to send is `SIGTERM`. I just want a nicer way to stop my child processes than `kill`, that's all. And I guess pretty many people who want to fiddle with child processes want to be able to kill them gracefully. I don't need a `send_signal` really, I just need `terminate`. Then it's good to expose a `send_signal` as well since it's no more work and it makes `duct` cater to many more special use cases.

A dedicated `terminate` let me do graceful kills without having to depend on `libc`. That might be a reason for it to exist?

**Comment by faern** [comment]
I believe one of your argument against `terminate` was that it was hard to explain to people what signals are, and hard to explain that it might not actually kill the child etc. I would argue that this is not our responsibility to fully explain. If the documentation clearly states that all it does is to send a `SIGTERM` and that procesess on unix based systems might not exit from this signal, then I think we explained enough and it's up to the consuming developer to research what that means and if it applies to their usecase.

**Comment by faern** [comment]
I'm also having some seconds thought about removing `then`. As a consumer of this library it is fully possible to manually do the same by having a thread waiting for the handle and then starting the next one etc, so it's not making something impossible by removing it. But the point is to provide convenient abstractions, right? It would require a fair amount of code and complexity to achieve the same as a user of `duct`.

If I look for a library that starts its documentation by saying "duct wants to make shelling out in Rust as easy and flexible as it is in Bash" then I expect it t support `foo && bar` and `foo; bar` :P

I would not say that implementing `send_signal` or `terminate` was that hard for `then`. It's just that it might have caveats. But as long as they are stated in the docs I think it's up to the consumer to decide if they want to use the feature or not and how they should use it in their use case. Just as an example: If they are only running processes that honor `SIGTERM`, great, then they can code with that assumption.

**Comment by faern** [comment]
Sorry for thinking out loud and for this wall of text.

I see two main categories of users who want to run child processes and thus might have use of `duct`:

1. People writing applications running fixed, known, expressions, doing a very limited thing in a very controlled manner. I fit in this category in my current use case. An example of this is someone who write bindings to btrfs (I don't), but they are too lazy to actually call the corresponding syscalls. So they just shell out to `btrfs subvolume create ...` and similar. The point is that they will know at the time they write the code what processes they will run and how these processes behave.

2. People writing intermediate libraries/applications that allow others on top of them execute processes. This includes people writing shells or any re-implementation of `watch` and similar commands.

My guess and generalization is that I think people in category 1 very rarely, or never, need `then` expressions, but they need graceful exits and they might need special signals to specific children. Category 2 know so little about what they will execute so they can't really make meaningful decisions about signals at the time of writing. However, they might want `then` expressions for convenience.

**Comment by oconnor663** [comment]
(I put this mentally on hold to finish factoring out the `sh` function. Will get back to it shortly.)

**Comment by faern** [comment]
@oconnor663 Ah, did not see your latest comment before I pushed these last commits. Now I reverted the `terminate` method commit. I still kept it in the history in case this discussion comes up again and we want to bring it back.

But as the PR stands now it adds the signaling functionality and it describes the `then` caveat in the documentation. Is this OK?

**Comment by faern** [comment]
Ping? :)

**Comment by oconnor663** [comment]
Blocking off some time this evening. Sorry for being a slowpoke!

**Comment by oconnor663** [comment]
Looks great! Merged and released as 0.9.1. Apologies again for making you wait so long.

I think my next step will be to take a shot at ripping out `then` and just seeing how that feels. (One of the questions in my head will be whether the tree structure of handles and expressions still makes sense. I think it will.)
